File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# Aditya Verma Approach
class Solution:
    def solve(self, root, res):
        # base condition
        if root is None:
            return 0
        
        # hypothesis
        left = self.solve(root.left, res)
        right = self.solve(root.right, res)
        
        # Intuation
        temp = max(left, right) + 1
        ans = max(temp, 1 + left + right)
        res[0] = max(res[0], ans)
        
        return temp
    
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        # use the res array res[0]
        res = [float('-inf')]
        self.solve(root, res)
        return res[0] - 1

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int diameterOfBinaryTree(TreeNode root) {
        int[] result = new int[]{Integer.MIN_VALUE}; // Mutable reference
        solve(root, result);
        return result[0]-1; // Returning diameter in terms of edges
    }

    private int solve(TreeNode root, int[] result) {
        if (root == null) return 0;

        int leftdia = solve(root.left, result);
        int rightdia = solve(root.right, result);

        int temp = 1 + Math.max(leftdia, rightdia);
        int ans = Math.max(temp,1+leftdia+rightdia); // Diameter is counted in edges (not nodes)
        result[0] = Math.max(result[0], ans); // Update global result

        return temp;
    }
}
